File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
       
        rows=len(grid)
        cols=len(grid[0])

        count=0
        visited=set()

        def bfs(r,c):
            directions=[[0,+1],[0,-1],[+1,0],[-1,0]]
            q=deque()
            q.append((r,c))
            while q:
                    ro,co=q.popleft()
                    for dr,dc in directions:
                        row=ro+dr
                        col=co+dc
                        if (row in range(rows) and col in range(cols) and grid[row][col]=='1' and (row,col) not in visited):
                            visited.add((row,col))
                            q.append((row,col))
        #---------------------------------

        for r in range(rows):
            for c in range(cols):
                if grid[r][c]=='1' and (r,c) not in visited:
                    bfs(r,c)
                    count+=1
                
        return count

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public void dfs(char[][] grid, int row,int col){
        if(row<0 || row>=grid.length || col<0 || col>=grid[0].length || grid[row][col]=='0'){
            return;
        }
        grid[row][col]='0';
        dfs(grid,row+1,col);
        dfs(grid,row-1,col);
        dfs(grid,row,col+1);
        dfs(grid,row,col-1);
    }
    public int numIslands(char[][] grid) {
        int m=grid.length;
        int n=grid[0].length;
        int count=0;
        for(int row=0;row<m;row++){
            for(int col=0;col<n;col++){
                if(grid[row][col]=='1'){
                    dfs(grid,row,col);
                    count++;
                }
            }
        }
        return count;
    }
}